RHIDP-2384: Clean-up Configuring the RHDH CR section & sub-sections (#232)
redhat-developer/red-hat-developers-documentation-rhdh: modules/installation/proc-add-custom-app-config-file-ocp-operator.adoc

> [id="proc-add-custom-app-config-file-ocp_{context}"]

Reviewer: ```suggestion
[id="proc-add-custom-app-config-file-ocp-operator_{context}"]
```
ID needs to match the filename

= Adding a custom application configuration file to {ocp-short} using the Operator

> A custom application configuration file is a ConfigMap that you can use to change the configuration of your {product} instance to meet the needs and preferences of your organization. If you are deploying your {product-short} instance on {ocp-brand-name}, you can use the {product} Operator to add a custom application configuration file to your {ocp-short} instance by creating the ConfigMap and referencing it in the {product-short} Custom Resource.

Reviewer: ```suggestion
A custom application configuration file is a `ConfigMap` object that you can use to change the configuration of your {product} instance. If you are deploying your {product-short} instance on {ocp-brand-name}, you can use the {product} Operator to add a custom application configuration file to your {ocp-short} instance, by creating the `ConfigMap` object and referencing it in the {product-short} custom resource (CR).
```
Couple of things:
- Suggested editing for minimalism to remove fluff in the first sentence.
- Suggested formatting of config map. Per the OCP guidelines and the RH supplementary style guide, for consistency across the org, we should either use the API object formatting if we're referring to a specific object ("`ConfigMap` object"), or it should be formatted as "config map".
- Suggested formatting of custom resource. Per the OCP guidelines and RHSG we should use "custom resource (CR)" for the first instance, and then you can use CR throughout the rest of the module etc if you like.

Links for ref:
- https://redhat-documentation.github.io/supplementary-style-guide/#custom-resource
- https://redhat-documentation.github.io/supplementary-style-guide/#config-map
- https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-formatting

> The custom application configuration file contains a sensitive environment variable, named `BACKEND_SECRET`, which is a mandatory backend authentication key that {product-short} uses to reference an environment variable defined in an {ocp-short} Secret. You must create a custom Secret, named 'secrets-rhdh', and reference it in the {product-short} Custom Resource.

Reviewer: ```suggestion
The custom application configuration file contains a sensitive environment variable named `BACKEND_SECRET`. This variable contains a mandatory backend authentication key that {product-short} uses to reference an environment variable defined in an {ocp-short} secret. You must create a secret named 'secrets-rhdh', and reference it in the {product-short} CR.
```

Suggestion to break up sentences so that this is easier to parse.
Similar to above comment, unless referring to a specific `Secret` object, this should be lowercase "secret" for OCP references.

[NOTE]
====
You are responsible for protecting your {product} installation from external and unauthorized access. Manage the backend authentication key like any other secret. Meet strong password requirements, do not expose it in any configuration files, and only inject it into configuration files as an environment variable.
====

.Prerequisites
* You have an active {ocp-brand-name} account.
* Your administrator has installed the {product} Operator in {ocp-short}. For more information, see link:{LinkAdminGuide}[Installing the {product} Operator].

> * You have created the {product} Custom Resource in {ocp-short}.

Reviewer: ```suggestion
* You have created the {product} CR in {ocp-short}.
```
Can we clarify which CR we're referring to here? Maybe throughout this doc rather than referring to it as the {product} CR, we can mention it by name and link to docs about creating it? Presumably it's maybe the `app-config.yaml`, but I'm not sure and it's not really clear?

.Procedure

> . From the *Developer* perspective, select the *ConfigMaps* tab.

Reviewer: ```suggestion
. From the *Developer* perspective in the {ocp-short} web console, select the *ConfigMaps* tab.
```
Should this be "view" instead of "tab", similar to Topology?

. Click *Create ConfigMap*.
. Select the *YAML view* option in *Configure via* and use the following example as a base template to create a ConfigMap, such as `app-config-rhdh.yaml`:
+
[source,yaml,subs="attributes+"]
----
kind: ConfigMap
apiVersion: v1
metadata:
  name: app-config-rhdh
data:
  "app-config-rhdh.yaml": |
    app:
      title: {product}
      baseUrl: https://backstage-<CUSTOM_RESOURCE_NAME>-<NAMESPACE_NAME>.<OPENSHIFT_INGRESS_DOMAIN>
    backend:
      auth:
        keys:
          - secret: "${BACKEND_SECRET}"
      baseUrl: https://backstage-<CUSTOM_RESOURCE_NAME>-<NAMESPACE_NAME>.<OPENSHIFT_INGRESS_DOMAIN>
      cors:
        origin: https://backstage-<CUSTOM_RESOURCE_NAME>-<NAMESPACE_NAME>.<OPENSHIFT_INGRESS_DOMAIN>

> ----

Reviewer: Per the Red Hat supplementary style guide, user-replaceable values should use `<this_format>`, so these should be lowercase throughout the doc. See https://redhat-documentation.github.io/supplementary-style-guide/#user-replaced-values

+

> .. Use the mandatory backend authentication key for {product} to reference an environment variable defined in an {ocp-short} Secret.

Reviewer: ```suggestion
.. Use the mandatory backend authentication key for {product} to reference an environment variable defined in an {ocp-short} secret.
```

> .. Set the external URL of your {product} instance in the `app.baseUrl`, `backend.baseUrl` and `backend.cors.origin` fields of the application configuration. By default, the URL is similar to the following example:

Reviewer: Same comment as above re https://redhat-documentation.github.io/supplementary-style-guide/#user-replaced-values being lowercase

`pass:c[https://backstage-<CUSTOM_RESOURCE_NAME>-<NAMESPACE_NAME>.<OPENSHIFT_INGRESS_DOMAIN>;]`.
+
[NOTE]
====

> You can use the `oc get ingresses.config/cluster -o jsonpath='{.spec.domain}'` command to display your ingress domain. If you want to use a different host or sub-domain, customize the Custom Resource `spec.application.route` field and adjust the application configuration accordingly.

Reviewer: ```suggestion
You can use the `oc get ingresses.config/cluster -o jsonpath='{.spec.domain}'` command to display your ingress domain. If you want to use a different host or sub-domain, customize the CR `spec.application.route` field and adjust the application configuration accordingly.
```

====

. Click *Create*.

> . Select the *Secrets* tab.

Reviewer: Same for all instances of tab - should these be "view" if they're clicked on from the same left nav as the Topology view?

. Click *Create Key/value Secret*.

> . Create a custom secret named `secrets-rhdh`.

Reviewer: ```suggestion
. Create a secret named `secrets-rhdh`.
```
I think "custom" is superfluous throughout if they're creating a new secret.

. Add a key named `BACKEND_SECRET` and a base64 encoded string as a value. Use a unique value for each {product} instance. For example, you can use the following command to generate a key from your terminal:
+
[source,yaml]
----
node -p 'require("crypto").randomBytes(24).toString("base64")'
----

. Click *Create*.
. Select the *Topology* view.
. Click the overflow menu for the {product} instance that you want to use and select *Edit Backstage* to load the YAML view of the {product} instance.
+
image::rhdh/operator-install-2.png[]

> . In the Custom Resource file, enter the name of the custom application configuration ConfigMap as the value for the `spec.application.appConfig.configMaps` field and the name of the custom Secret as the value for the `spec.application.extraEnvs.secrets` field. For example:

Reviewer: ```suggestion
. In the CR, enter the name of the custom application configuration config map as the value for the `spec.application.appConfig.configMaps` field, and enter the name of your secret as the value for the `spec.application.extraEnvs.secrets` field. For example:
```

+
[source, yaml]
----
spec:
  application:
    appConfig:
      mountPath: /opt/app-root/src
      configMaps:
         - name: app-config-rhdh
    extraEnvs:
      secrets:
         - name: secrets-rhdh
    extraFiles:
      mountPath: /opt/app-root/src
    replicas: 1
    route:
      enabled: true
  database:
    enableLocalDb: true

> ----

Reviewer: In this YAML example, I would include the API version, kind, etc so that it's clear what the custom resource is that's being modified here. For example, if it was a config map:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: example
spec:....
```

I think it's good practice to always include this for context for users.

The custom application configuration file contains a sensitive environment variable, named `BACKEND_SECRET`. This variable contains a mandatory backend authentication key that {product-short} uses to reference an environment variable defined in an {ocp-short} secret. You must create a secret, named 'secrets-rhdh', and reference it in the {product-short} CR.

[NOTE]
====
You are responsible for protecting your {product} installation from external and unauthorized access. Manage the backend authentication key like any other secret. Meet strong password requirements, do not expose it in any configuration files, and only inject it into configuration files as an environment variable.
====

.Prerequisites
* You have an active {ocp-brand-name} account.
* Your administrator has installed the {product} Operator in {ocp-short}. For more information, see link:{LinkAdminGuide}[Installing the {product} Operator].
* You have created the {product} CR in {ocp-short}.

.Procedure
. From the *Developer* perspective in the {ocp-short} web console, select the *ConfigMaps* view.
. Click *Create ConfigMap*.

> . Select the *YAML view* option in *Configure via* and use the following example as a base template to create a ConfigMap, such as `app-config-rhdh.yaml`:

Reviewer: ```suggestion
. Select the *YAML view* option in *Configure via* and use the following example as a base template to create a `ConfigMap` object, such as `app-config-rhdh.yaml`:
```
